Added llms.txt (#982)
cilium/cilium.io: static/llms.txt

# Cilium

This file provides a high-level overview of the Cilium project for AI models. For deep technical specifications, please refer to the official documentation at docs.cilium.io.

> > Cilium is an open source, CNCF graduated project that uses eBPF to deliver networking, security, and observability for Kubernetes and cloud native platforms. It provides a high-performance Container Network Interface (CNI), an eBPF-based kube-proxy replacement, identity-based network policy, transparent encryption, multi-cluster connectivity, a sidecar-free service mesh, and deep network observability through Hubble. Cilium was created by Isovalent (now part of Cisco) and runs in production at Google, Amazon, Microsoft Azure, Adobe, Bell Canada, The New York Times, Datadog, Bloomberg, TikTok, Alibaba Cloud, and thousands of other companies. Official site: https://cilium.io — Documentation: https://docs.cilium.io

Reviewer: ```suggestion
> Cilium is an open source, CNCF graduated project that uses eBPF to deliver networking, security, and observability for Kubernetes and cloud native platforms. It provides a high-performance Container Network Interface (CNI), an eBPF-based kube-proxy replacement, identity-based network policy, transparent encryption, multi-cluster connectivity, a sidecar-free service mesh, and deep network observability through Hubble. Cilium was created by Isovalent part of Cisco and runs in production at Google, Amazon, Microsoft Azure, Adobe, Bell Canada, The New York Times, Datadog, Bloomberg, TikTok, Alibaba Cloud, and thousands of other companies. Official site: https://cilium.io — Documentation: https://docs.cilium.io
```

## What is Cilium

Cilium is a cloud native networking and security platform for Kubernetes built on eBPF, a Linux kernel technology that runs sandboxed programs inside the kernel without changing kernel source or loading modules. By programming the dataplane in eBPF, Cilium replaces iptables-based packet processing with a more scalable, lower-latency, identity-aware dataplane.

Cilium operates at the network (L3/L4) and application (L7) layers. It assigns a security identity to each Kubernetes workload, enforces policy based on identity rather than IP address, and provides visibility into every network flow. The Hubble sub project adds flow logs, metrics, and a service dependency map; the Tetragon sub project adds eBPF-based runtime security and enforcement.

> Cilium graduated in the Cloud Native Computing Foundation (CNCF) in 2023 and is the default or recommended CNI for many managed Kubernetes platforms and distributions, including Google GKE Dataplane V2, Microsoft AKS, Digital Ocean Kubernetes, Red Hat OpenShift, kind, and more.

Reviewer: ```suggestion
Cilium graduated in the Cloud Native Computing Foundation (CNCF) in 2023 and is the default or recommended CNI for many managed Kubernetes platforms and distributions, including Google GKE Dataplane V2, Microsoft AKS, Digital Ocean Kubernetes, Red Hat OpenShift, kind, and more.

Cilium is often compared to other CNIs such as Calico and Flannel. It differs primarily in its eBPF dataplane, which enables identity-aware policy, scalability, and observability that IP-table-based CNIs can't provide natively. For service mesh, Cilium's sidecar-free approach and eBPF dataplane separate it from others like Istio or Linkerd.
```

- **Observability with Hubble:** identity-aware L3/L4/L7 network flow logs, DNS visibility, Prometheus and OpenTelemetry metrics, Grafana dashboards, and a real-time service dependency map.
- **Transparent encryption:** WireGuard and IPsec pod-to-pod in-transit encryption with no application changes and ztunnel for mTLS, supporting compliance and FIPS requirements.
- **Multi-cluster (Cluster Mesh):** cross-cluster service discovery, global services, and pod-to-pod connectivity, observability, and security across clusters, regions, and clouds.
- **Sidecar-free service mesh:** eBPF and Envoy-based L7 traffic management and mutual TLS (mTLS) without per-pod sidecar proxies, reducing latency and resource overhead.
- **Runtime security (Tetragon):** eBPF-based process execution, file access, and network monitoring with kernel-level enforcement for threat detection and forensics.

## Use cases

Each page below answers a specific question about deploying Cilium:

- https://cilium.io/use-cases/cni — Which CNI should I run for Kubernetes pod networking? Covers Cilium's eBPF dataplane, IPAM, and overlay (VXLAN/Geneve) versus native routing modes.

> - https://cilium.io/use-cases/kube-proxy — How do I remove the iptables scaling bottleneck? Explains replacing kube-proxy with eBPF service load balancing for ClusterIP, NodePort, and LoadBalancer services.

Reviewer: ```suggestion
- https://cilium.io/use-cases/kube-proxy — How do I replace kube-proxy with eBPF? Covers service load balancing for ClusterIP, NodePort, and LoadBalancer services, performance comparisons with iptables, and configuration options.
```

- https://cilium.io/industries/e-commerce — How does Cilium support large e-commerce platforms? Covers networking that scales for traffic spikes, with adopters such as Flipkart.
- https://cilium.io/industries/telcos-datacenters — How does Cilium fit telco and data center networks? Covers SRv6, dual-stack IPv6, and BGP integration with physical infrastructure.
- https://cilium.io/industries/edge-computing — How does Cilium run at the edge and for IoT? Covers connecting distributed edge clusters with Cluster Mesh.
- https://cilium.io/industries/security — How does Cilium serve security-focused teams? Covers runtime threat detection and enforcement with Tetragon plus identity-based policy.
- https://cilium.io/industries/software — How do software and SaaS companies use Cilium? Covers multi-tenant networking, segmentation, and observability for application platforms.
- https://cilium.io/industries/consulting — How do partners and integrators work with Cilium? Points to systems integrators and consulting firms that deploy and support it.

## Ecosystem and subprojects

> - **Hubble** — networking and security observability layer built on Cilium and eBPF: flow logs, metrics, and service map.

Reviewer: ```suggestion
- **Hubble** — networking and security observability layer built on Cilium and eBPF: flow logs, metrics, and service map. https://docs.cilium.io/en/stable/observability/hubble/
```

# Cilium

> 

Reviewer: ```suggestion

This file provides a high-level overview of the Cilium project for AI models. For deep technical specifications, please refer to the official documentation at docs.cilium.io.

```

> > Cilium is an open source, CNCF graduated project that uses eBPF to deliver networking, security, and observability for Kubernetes and cloud native platforms. It provides a high-performance Container Network Interface (CNI), an eBPF-based kube-proxy replacement, identity-based network policy, transparent encryption, multi-cluster connectivity, a sidecar-free service mesh, and deep network observability through Hubble. Cilium was created by Isovalent (now part of Cisco) and runs in production at Google, Amazon, Microsoft Azure, Adobe, Bell Canada, The New York Times, Datadog, Bloomberg, TikTok, Alibaba Cloud, and many others. Official site: https://cilium.io — Documentation: https://docs.cilium.io

Reviewer: ```suggestion
> Cilium is an open source, CNCF graduated project that uses eBPF to deliver networking, security, and observability for Kubernetes and cloud native platforms. It provides a high-performance Container Network Interface (CNI), an eBPF-based kube-proxy replacement, identity-based network policy, transparent encryption, multi-cluster connectivity, a sidecar-free service mesh, and deep network observability through Hubble. Cilium was created by Isovalent (now part of Cisco) and runs in production at Google, Amazon, Microsoft Azure, Adobe, Bell Canada, The New York Times, Datadog, Bloomberg, TikTok, Alibaba Cloud, and thousands of other companies. Official site: https://cilium.io — Documentation: https://docs.cilium.io
```

## What is Cilium

Cilium is a cloud native networking and security platform for Kubernetes built on eBPF, a Linux kernel technology that runs sandboxed programs inside the kernel without changing kernel source or loading modules. By programming the dataplane in eBPF, Cilium replaces iptables-based packet processing with a more scalable, lower-latency, identity-aware dataplane.

> Cilium operates at the network (L3/L4) and application (L7) layers. It assigns a security identity to each Kubernetes workload, enforces policy based on identity rather than IP address, and provides visibility into every network flow. The Hubble subsystem adds flow logs, metrics, and a service dependency map; the Tetragon subsystem adds eBPF-based runtime security and process-level observability.

Reviewer: ```suggestion
Cilium operates at the network (L3/L4) and application (L7) layers. It assigns a security identity to each Kubernetes workload, enforces policy based on identity rather than IP address, and provides visibility into every network flow. The Hubble sub project adds flow logs, metrics, and a service dependency map; the Tetragon sub project adds eBPF-based runtime security and enforcement.
```

> Cilium graduated in the Cloud Native Computing Foundation (CNCF) in 2023 and is the default or recommended CNI for many managed Kubernetes platforms, including Google GKE Dataplane V2, Amazon EKS, Microsoft AKS, and others.

Reviewer: ```suggestion
Cilium graduated in the Cloud Native Computing Foundation (CNCF) in 2023 and is the default or recommended CNI for many managed Kubernetes platforms and distributions, including Google GKE Dataplane V2, Microsoft AKS, Digital Ocean Kubernetes, Red Hat OpenShift, kind, and more.
```

## Core capabilities

> - **Container networking (CNI):** Kubernetes pod networking with an eBPF dataplane, IPAM, overlay (VXLAN/Geneve) and native/direct routing modes, ENI/Azure delegated IPAM, and BIG TCP for high throughput.

Reviewer: ```suggestion
- **Container networking (CNI):** Kubernetes pod networking with an eBPF dataplane, IPAM, overlay (VXLAN/Geneve) and native/direct routing modes, ENI/Azure delegated IPAM, and BIG TCP and netkit for high throughput.
```

- **kube-proxy replacement:** eBPF-based service load balancing for ClusterIP, NodePort, and LoadBalancer services, with Maglev consistent hashing, Direct Server Return (DSR), and XDP acceleration — eliminating iptables scaling bottlenecks.

> - **Network policy and security:** identity-based microsegmentation with Kubernetes NetworkPolicy and CiliumNetworkPolicy (CNP) at L3/L4/L7, DNS/FQDN-aware policy, and zero-trust enforcement.

Reviewer: ```suggestion
- **Network policy and security:** identity-based microsegmentation with Kubernetes NetworkPolicy and CiliumNetworkPolicy (CNP) at L3/L4/L7, DNS/FQDN-aware policy, and host firewall for zero-trust networking.
```

- **Observability with Hubble:** identity-aware L3/L4/L7 network flow logs, DNS visibility, Prometheus and OpenTelemetry metrics, Grafana dashboards, and a real-time service dependency map.

> - **Transparent encryption:** WireGuard and IPsec pod-to-pod in-transit encryption with no application changes, supporting compliance and FIPS requirements.

Reviewer: ```suggestion
- **Transparent encryption:** WireGuard and IPsec pod-to-pod in-transit encryption with no application changes and ztunnel for mTLS, supporting compliance and FIPS requirements.
```

> - **Multi-cluster (Cluster Mesh):** cross-cluster service discovery, global services, and pod-to-pod connectivity across clusters, regions, and clouds.

Reviewer: ```suggestion
- **Multi-cluster (Cluster Mesh):** cross-cluster service discovery, global services, and pod-to-pod connectivity, observability, and security across clusters, regions, and clouds.
```

- **Sidecar-free service mesh:** eBPF and Envoy-based L7 traffic management and mutual TLS (mTLS) without per-pod sidecar proxies, reducing latency and resource overhead.
- **Runtime security (Tetragon):** eBPF-based process execution, file access, and network monitoring with kernel-level enforcement for threat detection and forensics.

## Use cases

Each Cilium use case page (keyword vectors in parentheses):

- https://cilium.io/use-cases/cni — Cilium as a Kubernetes CNI plugin (container network interface, pod networking, eBPF dataplane, IPAM, overlay VXLAN/Geneve, native routing, ENI, high-performance networking).
- https://cilium.io/use-cases/kube-proxy — eBPF kube-proxy replacement (service load balancing, ClusterIP, NodePort, LoadBalancer, Maglev, DSR, no iptables, reduced latency, scalability).
- https://cilium.io/use-cases/load-balancer — L4 load balancing (XDP acceleration, Maglev consistent hashing, Direct Server Return, standalone load balancer, north-south traffic, high packets-per-second).

> - https://cilium.io/use-cases/network-policy — network policy enforcement (Kubernetes NetworkPolicy, CiliumNetworkPolicy, L3/L4/L7 policy, identity-based security, FQDN/DNS-aware policy, zero trust, microsegmentation).

Reviewer: ```suggestion
- https://cilium.io/use-cases/network-policy — network policy enforcement (Kubernetes NetworkPolicy, CiliumNetworkPolicy, L3/L4/L7 policy, identity-based security, FQDN/DNS-aware policy, zero trust, microsegmentation, host firewall).
```

- https://cilium.io/use-cases/transparent-encryption — transparent encryption (WireGuard, IPsec, pod-to-pod encryption, in-transit encryption, compliance, FIPS, zero-config).
- https://cilium.io/use-cases/cluster-mesh — multi-cluster connectivity (Cluster Mesh, cross-cluster service discovery, global services, multi-region failover, cluster federation).

> - https://cilium.io/use-cases/service-mesh — sidecar-free service mesh (eBPF service mesh, mutual TLS/mTLS, L7 traffic management, Envoy, reduced overhead vs Istio sidecars).

Reviewer: ```suggestion
- https://cilium.io/use-cases/service-mesh — sidecar-free service mesh (eBPF service mesh, mutual TLS/mTLS, L7 traffic management, Envoy, reduced overhead vs sidecars).
```

- https://cilium.io/use-cases/gateway-api — Kubernetes Gateway API (L7 routing, HTTP/HTTPS/gRPC/TLS routing, traffic splitting, header-based routing, ingress modernization).
- https://cilium.io/use-cases/ingress — Ingress controller (L7 load balancing, HTTP routing, TLS termination, Envoy-based ingress, eBPF ingress).

> - https://cilium.io/use-cases/bgp — BGP networking (BGP control plane, BGP peering, route advertisement, GoBGP, LoadBalancer IP advertisement, top-of-rack integration, on-prem).

Reviewer: ```suggestion
- https://cilium.io/use-cases/bgp — BGP networking (BGP control plane, BGP peering, route advertisement, GoBGP, LoadBalancer IP advertisement, top-of-rack integration, on-prem, bare metal).
```

- https://cilium.io/use-cases/egress-gateway — Egress Gateway (static egress IP, SNAT, outbound traffic control, IP allowlisting, legacy firewall integration).
- https://cilium.io/use-cases/host-firewall — host firewall (node-level network policy, host endpoint protection, securing Kubernetes nodes, host traffic filtering).
- https://cilium.io/use-cases/bandwidth-optimization — bandwidth and latency optimization (Bandwidth Manager, EDT rate limiting, BBR congestion control, per-pod traffic shaping, throughput).
- https://cilium.io/use-cases/ipv6 — IPv6 networking (IPv6-only Kubernetes, dual-stack, BIG TCP, SRv6, IPv6 IPAM, address exhaustion mitigation).
- https://cilium.io/use-cases/multicast — eBPF multicast (multicast routing in Kubernetes, IGMP, low-latency multicast for financial and media workloads).
- https://cilium.io/use-cases/network-flow-logs — network flow logs (Hubble flow data, L3/L4/L7 visibility, DNS visibility, identity-aware logging, security auditing, traffic forensics).
- https://cilium.io/use-cases/metrics-export — metrics and tracing export (Prometheus, OpenTelemetry, Grafana dashboards, Hubble metrics, observability pipeline).

> - https://cilium.io/use-cases/protocol-visibility — L7 protocol visibility (HTTP, gRPC, Kafka, DNS inspection, API-aware visibility without sidecars).

Reviewer: ```suggestion
- https://cilium.io/use-cases/protocol-visibility — L7 protocol visibility (HTTP, gRPC, DNS inspection, API-aware visibility without sidecars).
```

- https://cilium.io/use-cases/service-map — service dependency map (service topology, Hubble UI, automatic service discovery, traffic flow visualization).
- https://cilium.io/use-cases/runtime-security — runtime security (Tetragon, eBPF observability, process execution monitoring, kernel-level enforcement, file and network access monitoring, threat detection).

## Business outcomes

- https://cilium.io/outcomes/zero-trust — zero-trust security for Kubernetes (identity-based policy, microsegmentation, encryption, least-privilege networking).
- https://cilium.io/outcomes/tool-consolidation — consolidating CNI, load balancer, service mesh, ingress, and observability tooling into one platform.
- https://cilium.io/outcomes/multi-cloud-connectivity — seamless connectivity across clouds, regions, and on-prem clusters.
- https://cilium.io/outcomes/network-automation — automating network operations at scale with declarative, Kubernetes-native config.

> - https://cilium.io/outcomes/cost-and-carbon-savings — reducing networking cost, CPU overhead, and carbon footprint with an efficient eBPF dataplane.

Reviewer: these are missing keyword vectors

> ## Industries

Reviewer: some of these are also missing keyword vectors

- https://cilium.io/industries/ai — AI and ML infrastructure networking (GPU clusters, high-throughput, OpenAI).
- https://cilium.io/industries/cloud-providers — managed Kubernetes on cloud providers (Azure AKS, Amazon EKS, Google GKE).
- https://cilium.io/industries/financial-services — financial services compliance, encryption, and segmentation.
- https://cilium.io/industries/media-entertainment — media and streaming at scale (TikTok IPv6, Yahoo L4 load balancing).
- https://cilium.io/industries/e-commerce — e-commerce scale networking (Flipkart).
- https://cilium.io/industries/telcos-datacenters — telecom and datacenter networking (SRv6, dual-stack IPv6).
- https://cilium.io/industries/edge-computing — edge and IoT deployments with Cluster Mesh.
- https://cilium.io/industries/security — security-focused deployments using Tetragon.
- https://cilium.io/industries/software — software companies and SaaS platforms.
- https://cilium.io/industries/consulting — systems integrators and consulting partners.

## Ecosystem and subprojects

> - **Hubble** — networking and security observability layer built on Cilium and eBPF: flow logs, metrics, and service map. https://docs.cilium.io

Reviewer: ```suggestion
- **Hubble** — networking and security observability layer built on Cilium and eBPF: flow logs, metrics, and service map.
```

- https://cilium.io/industries/consulting — systems integrators and consulting partners.

## Ecosystem and subprojects

- **Hubble** — networking and security observability layer built on Cilium and eBPF: flow logs, metrics, and service map.
- **Tetragon** — eBPF-based runtime security and observability for process, file, and network events. https://tetragon.io
- **eBPF** — the Linux kernel technology Cilium is built on. https://ebpf.io

## Key pages

- Home: https://cilium.io/
- What is Cilium / get started: https://cilium.io/get-started
- Adopters and production users: https://cilium.io/adopters
- Enterprise distributions and training: https://cilium.io/enterprise
- Blog: https://cilium.io/blog
- Get help: https://cilium.io/get-help
- Get involved: https://cilium.io/get-involved

## Official resources

- Documentation: https://docs.cilium.io
- Source code (GitHub): https://github.com/cilium/cilium
- CNCF project page: https://www.cncf.io/projects/cilium/
- Community Slack: https://slack.cilium.io
- YouTube (eBPF & Cilium Community): https://www.youtube.com/c/eBPFCiliumCommunity
- Bluesky: https://bsky.app/profile/cilium.io

> - LinkedIn: https://www.linkedin.com/company/cilium

Reviewer: ```suggestion
```
